### PR TITLE
librados: switch to use symbol versioning

### PR DESCRIPTION
**Pull request by dillaman**
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by dillaman** [comment]
@tchaikov This is a very preliminary start to adding explicit symbol versions to all librados API methods. The librados2 C API is nearly finished. The ```rados_tmap_XYZ``` and ```rados_ioctx_pool_stat``` show how it works around the broken ABI under nautilus.

The C++ is still a work-in-progress since I need to wrap so many methods (>200 ```ceph::buffer``` and >300 ```librados``` public methods). 

The 14.2.0 version of ```ceph::buffer::list``` from the API is hidden by dummy API wrappers to facilitate the necessary symbol aliasing. The pre-Nautilus backwards compatibly version of ```ceph::buffer::list``` (to address the internal structure changes) will just re-implement the older list implementation in a private implementation. 

The C++ librados API methods still need to be versioned and any methods using ```ceph::buffer::list``` need to have an extra backwards compatibility version. I will need to create conversion wrappers to switch back-and-forth between the two non-ABI compatible ```ceph::buffer::list``` formats within those methods.

**Comment by dillaman** [comment]
Cleanest RBD runs in months -- now real failures are visible:
http://pulpito.ceph.com/jdillaman-2019-02-05_15:17:05-rbd-wip-jd-testing-distro-basic-smithi/
http://pulpito.ceph.com/jdillaman-2019-02-05_19:04:43-rados-wip-jd-testing-distro-basic-smithi/

**Comment by smithfarm** [comment]
@dillaman Add `Fixes: https://tracker.ceph.com/issues/38177` to the commit message of the last commit?

**Comment by dillaman** [comment]
jenkins render docs

**Comment by tchaikov** [comment]
@dillaman thanks a lot for fixing this the right way. i admit that my librados2-compact was cheating =(. 

will take a closer look at this PR late tomorrow.

File: src/librados/librados_c.cc [hunk 1]
```diff
@@ -91,7 +99,7 @@ static CephContext *rados_create_cct(const char * const clustername,
   return cct;
 }
 
-extern "C" int rados_create(rados_t *pcluster, const char * const id)
+extern "C" int __rados_create(rados_t *pcluster, const char * const id)
```
Comment (tchaikov) [inline src/librados/librados_c.cc]: in C standard, 

> All identifiers that begin with an underscore and either an uppercase letter or another underscore are always reserved for any use.

see https://www.gnu.org/software/libc/manual/html_node/Reserved-Names.html

File: src/librados/librados_cxx_buffer.cc [hunk 2]
```diff
@@ -0,0 +1,204 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+/// duplicate ceph::buffer implementation within this unit
+#include "common/buffer.cc"
+#include "librados/librados_cxx.h"
+
+/// symbol versioning helpers for ceph::buffer
+#define CEPH_BUFFER_NS_API(fn)                            \
+  LIBRADOS_CXX_API(_ZN4ceph6buffer, fn)
```
Comment (chardan) [inline src/librados/librados_cxx_buffer.cc]: I'm a bit unclear on how this expected to work. The C++ standard does not guarantee name mangling conventions besides that required for C linkage. So... where does this leave us with compiling for clang, etc.? If the problem being solved is symbol exporting for C, maybe the way to do it is with a wrapper module?

Comment (dillaman) [reply src/librados/librados_cxx_buffer.cc]: GCC and Clang both follow the IA64 C++ mangling conventions [1]. For environments that don't support it, the fallback would be to just export everything w/o symbol versioning.


[1] https://itanium-cxx-abi.github.io/cxx-abi/

Comment (tchaikov) [reply src/librados/librados_cxx_buffer.cc]: correct. C++ standard does not define the name mangling convention. but in practice, the mainstream compilers like GCC, Clang, Intel C++ compiler all use the IA64 C++ ABI Spec. see https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling, which is well defined.

Comment (tchaikov) [reply src/librados/librados_cxx_buffer.cc]: and see https://github.com/gcc-mirror/gcc/blob/master/gcc/cp/mangle.c

Comment (chardan) [reply src/librados/librados_cxx_buffer.cc]: Agreed that the convention is relatively stable:
https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html

Note that name mangling can change across architectures-- when sizeof(int) changes. GCC for instance has changed before. Is it possible to accomplish this with inline namespaces in our case? If not, ok, at least we're doing it with open eyes! :-)

Comment (dillaman) [reply src/librados/librados_cxx_buffer.cc]: Hmm -- that inline namespace idea actually sounds much cleaner. Thanks.

Comment (dillaman) [reply src/librados/librados_cxx_buffer.cc]: ... with the apparent side-effect that all the existing forward declarations to librados C++ classes need to be yanked or wrapped with a version-specific inline. 

e.g. lots of code does things like "namespace librados { class AioCompletion; }" which make AioCompletion ambiguous when its moved to an inline namespace under librados.

Comment (chardan) [reply src/librados/librados_cxx_buffer.cc]: If I'm following your meaning, then yeah there would have to be some revision. I've used the feature in a basic context:
namespace foo { inline namespace version1 { ... } } 
But I haven't had opportunity to really consider the cases of reopening namespaces and understanding how that interacts with this feature yet. Looks like we might get to! Hope that helps.

File: src/librados/librados_c.cc (hunk 1 above)
Comment (dillaman) [reply src/librados/librados_c.cc]: Fixed

File: src/librados/librados_cxx_buffer.cc (hunk 2 above)
Comment (dillaman) [reply src/librados/librados_cxx_buffer.cc]: I added a librados forward declarations header file and yanked the individual forward declarations throughout. I think this inline namespace technique is a much nicer path forward.

**Comment by dillaman** [comment]
jenkins test docs

**Comment by dillaman** [comment]
retest this please

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2019-02-20_11:42:13-rbd-wip-jd-testing-distro-basic-smithi/
http://pulpito.ceph.com/jdillaman-2019-02-20_12:34:30-rados-wip-jd-testing-distro-basic-smithi/
http://pulpito.ceph.com/jdillaman-2019-02-20_12:39:36-rgw-wip-jd-testing-distro-basic-smithi/
http://pulpito.ceph.com/jdillaman-2019-02-20_12:41:58-fs-wip-jd-testing-distro-basic-smithi/

**Comment by dillaman** [comment]
Few initial failures in rados tests:
test_librados_build.sh: missed ```-lradospp``` -> ```-lrados```
test_envlibrados_for_rocksdb.sh: reverted patch to switch to libradospp
rados bench crashed due to unrelated buffer::list issues that have since been reverted in master

**Comment by dillaman** [comment]
jenkins test docs

**Comment by dillaman** [comment]
RGW/FS re-run due to cpeh-mgr-ssh issue:
http://pulpito.ceph.com/jdillaman-2019-02-20_17:49:07-rgw-wip-jd-testing-distro-basic-smithi/
http://pulpito.ceph.com/jdillaman-2019-02-20_17:49:52-fs-wip-jd-testing-distro-basic-smithi/

File: src/cls/rbd/cls_rbd.cc [hunk 3]
```diff
@@ -3177,7 +3177,7 @@ int dir_state_assert(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
     return -EINVAL;
   }
 
-  cls::rbd::DirectoryState on_disk_directory_state;
+  cls::rbd::DirectoryState on_disk_directory_state = directory_state;
```
Comment (tchaikov) [inline src/cls/rbd/cls_rbd.cc]: @dillaman is this related to the versioning of C++ API? if not, can we move it into its own commit?

Comment (dillaman) [reply src/cls/rbd/cls_rbd.cc]: Done
